server/src/main/java/io/crate/execution/engine/collect/files/LineParser.java

```diff
@@ -44,25 +47,39 @@ public LineParser(CopyFromParserProperties parserProperties, List<String> target
         this.targetColumns = targetColumns;
     }
 
-    private enum InputType {
+    public enum InputType {
         CSV,
         JSON
     }
 
-    public void readFirstLine(URI currentUri,
-                              FileUriCollectPhase.InputFormat inputFormat,
-                              BufferedReader currentReader) throws IOException {
+    public InputType inputType() {
+        return inputType;
+    }
+
+    /**
+     * @return all actual columns (including non-existing in the target table) so that new columns can be added to the target columns
+     * or NULL if no need to update target columns.
+     */
+    @Nullable
+    public Collection<String> readFirstLine(URI currentUri,
+                                            FileUriCollectPhase.InputFormat inputFormat,
+                                            BufferedReader currentReader) throws IOException {
         for (long i = 0; i < parserProperties.skipNumLines(); i++) {
             currentReader.readLine();
         }
         if (isInputCsv(inputFormat, currentUri)) {
             csvLineParser = new CSVLineParser(parserProperties, targetColumns);
+            inputType = InputType.CSV;
             if (parserProperties.fileHeader()) {
-                csvLineParser.parseHeader(currentReader.readLine());
+                return Arrays.asList(csvLineParser.parseHeader(currentReader.readLine()));
+            } else {
```

Comment:
This happens only once as it's CSV specific so fine to wrap into some collection. It's done as a favour towards more heavy per row JSON which operates with raw key set --> collection.